### PR TITLE
CB-18539 Cannot start termination flow if the DBStack is in any delet…

### PR DESCRIPTION
**Pull request by topolyai5**
…ed phase but the stack doesn't have any active termination flow

See detailed description in the commit message.

### CODE REVIEW

**Review by gergopapi2 — CHANGES_REQUESTED** [review]

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsTerminationService.java [hunk 1]
```diff
@@ -76,9 +85,16 @@ private DatabaseServerConfig terminate(DatabaseServerConfig server, boolean forc
         return server;
     }
 
-    private boolean deletetedOnProviderSide(DatabaseServerConfig server, DBStack dbStack) {
-        // This is the case when user only delete the database on the provider but not the Cloudformation template
-        return !server.isArchived() && dbStack.getStatus().isDeleteCompleted();
+    @VisibleForTesting
+    boolean isTerminatedFlowAlreadyRun(Long id) {
```
[the remark this answers is not in the document]
Comment (topolyai5) [reply redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsTerminationService.java]: Good idea, I have updated

File: flow/src/main/java/com/sequenceiq/flow/service/flowlog/FlowLogDBService.java [hunk 2]
```diff
@@ -333,6 +334,17 @@ public List<FlowLog> findAllByFlowIdOrderByCreatedDesc(String flowId) {
         return flowLogRepository.findAllByFlowIdOrderByCreatedDesc(flowId);
     }
 
+    public boolean isFlowConfigAlreadyRun(Long id, Class<? extends FlowConfiguration<?>> flowConfigurationClass) {
```
Comment (gergopapi2) [inline flow/src/main/java/com/sequenceiq/flow/service/flowlog/FlowLogDBService.java:337]: I apologize for one more comment: I would change the method name. I think the name isFlowConfigAlreadyRunnING expresses the goal of the method better. AlreadyRun for me means that the flow ran at some point in the past and could be finished already.

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsTerminationService.java [hunk 3]
```diff
@@ -61,7 +66,9 @@ private DatabaseServerConfig terminate(DatabaseServerConfig server, boolean forc
         if (LOGGER.isDebugEnabled()) {
             LOGGER.debug("Terminate called for: {} with force: {}", dbStack, force);
         }
-        if (dbStack.getStatus().isDeleteInProgressOrCompleted() && !deletetedOnProviderSide(server, dbStack)) {
+
+        if (flowLogService.isFlowConfigAlreadyRun(dbStack.getId(), RedbeamsTerminationFlowConfig.class)
+                || (server.isArchived() && dbStack.getStatus().isDeleteCompleted())) {
```
Comment (pdarvasi) [inline redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsTerminationService.java:71]: why have you modified the second clause, the private method was more descriptive, what's the goal of removing it?

File: flow/src/main/java/com/sequenceiq/flow/service/flowlog/FlowLogDBService.java (hunk 2 above)
Comment (pdarvasi) [reply flow/src/main/java/com/sequenceiq/flow/service/flowlog/FlowLogDBService.java:337]: +1

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsTerminationService.java (hunk 3 above)
Comment (topolyai5) [reply redbeams/src/main/java/com/sequenceiq/redbeams/service/stack/RedbeamsTerminationService.java:71]: the logic is changed there and the name isn't matched for the logic. Here we should skip the deletion if archived and delete completed. The method was this: it is not archived and deleted. However, I can change the method name to already deleted or smth like this

**Review by pdarvasi — APPROVED** [review]

**Comment by topolyai5** [comment]
@gergopapi2 could you please check it again?

File: flow/src/test/java/com/sequenceiq/flow/service/flowlog/FlowLogDBServiceTest.java [hunk 4]
```diff
@@ -338,6 +340,45 @@ private FlowLog createFlowLog(String flowId) {
         return flowLog;
     }
 
+    @Test
+    void testIsTerminatedFlowAlreadyRunWhenHasActiveTerminationFlow() throws ClassNotFoundException {
```
Comment (gergopapi2) [inline flow/src/test/java/com/sequenceiq/flow/service/flowlog/FlowLogDBServiceTest.java:344]: nit: alreadyRun => alreadyRunning

**Review by gergopapi2 — APPROVED** [review]

**Pull request merged**